### DISCUSSION TITLE
How to get rawXmlString

### DISCUSSION THREAD

**Discussion by martijn10kb** [dolanmiu/docx#1933] [Q&A] [answered]
Heey,
I need to get an XML string of something I have build with this module.
I am at the  point where I have an array of elements, and need to convert that to xml so I can then insert that into an already existing docx (no worries I know this module doesn't do that..) just need to xml string.

We currently have a docx document as template in which we need some rich text elements to be inserted.
Now we already have a HTML to "array with 'docx' elements" script. Just need a way to convert that array to a raw XML string so I can do a search and replace in my existing docx with that resulting XML string.

**Comment by dolanmiu** [comment, marked as answer]
Currently that is not exposed, but that is a good idea, perhaps it can be part of `Packer`

However, I am actually making a solution right now to allow for documents to be edited, by exactly the method you are describing. No ETA yet, but it will be in the next release.

**Reply by martijn10kb** [reply]
Cool, sounds very promising!
When you have an ETA, or when it's released. Could you notify me? :)

**Reply by dolanmiu** [reply]
Will keep you updated!

**Reply by martijn10kb** [reply]
@dolanmiu Wait what you're working on, does it also involve the search and replace part? If so, can it also read the current style of what is to be replaced? And then apply that to the elements that are inserted?

**Reply by dolanmiu** [reply]
Yes, search and replace, thinking of using a handle bars type notation

```
{{ MY_COOL_TEXT }}
```

Then in the TypeScript:

```ts
const data = fs.readFileSync("path/to/file.docx");

await patchDocument(data, {
  patches: [{
    id: 'MY_COOL_TEXT',
    children: [new Paragraph(''), new Paragraph('')]
  }]
});
```

Then maybe images can be done later. Images are slightly more complex because it needs the relationship id too etc
Something like this

**Reply by martijn10kb** [reply]
Sounds good!

Wat I currently have in my docx is:
```
{{ convertHtmlToXml(data.standard_texts.h01_heading.value) }}
{{ convertHtmlToXml(data.standard_texts.h01_first_paragraph.value) }}
```
and then in my code I have a `data` object that contains.
```
const data = {
  standard_texts: {
    h01_heading: { id: 312, value: '<p>foo bar buzz</p>', title: 'some about foo' },
    h01_paragraph: { id: 312, value: '<p>Foo Fighters is an American rockband</p>', title: 'first paragraph about foo' },
    ...
  },
  ...
}
const convertHtmlToXml = (html: string) => {
  // do 'docx' magic
}
```
Although I do think now, that it shouldn't convert to XML then, but just to docx elements, and then use you code to insert that.

btw, the other moduel I'm using to interpolate docx is 'docx-templates'.. but does not and will not fully cover my needs..

**Reply by dolanmiu** [reply]
Here is an update:

https://github.com/dolanmiu/docx/pull/1960

API will look something like this:

https://github.com/dolanmiu/docx/blob/8c0fe00c6f9f9edd5d4836c5db865372062fe6c8/demo/85-template-document.ts

**Reply by ewoutkleinsmann** [reply]
Really nice!

I just tested it out and noticed that the children in the patches can only be TextRuns, right? Do you plan to also support Paragraphs and Tables for example?

**Reply by dolanmiu** [reply]
Yes, planning to support Paragraph and Tables too, working on that next.

Will need to think of how the replacement works

E.g.

```
Hello, the weather is {{status}} today.
```

If a paragraph replacement is specified for {{ good }}, then the entire paragraph, (and thus that sentence) is replaced. Unsure if this would confuse the user or not

After that, then images will be next

**Reply by martijn10kb** [reply]
Ohh looks really promising.
 
 I think what we need to have to move forward for our project is `ImageRun`, `ExternalHyperlink`, `TextRun`, and `Paragraph`. And the ability that we can read the current style of the document. So that in your above example the `status` will be in the same font style as where `{{status}}` (and the rest of the paragraph/textrun) is written in.
 
 The other module we use is [docx-templates](https://github.com/guigrpa/docx-templates) how to have it set up is that he use can choose the `delimiters` to be `{{`&`}}` so you get:
 ```
 ipsum lorem {{ HTML foo.bar }} sit amet.
 ```
 then the string in `foo.bar` will receive an altChunk in the document.
 
 ps. to avoid confusement. Ewout is a colleague of my.

**Reply by dolanmiu** [reply]
Images are now working, almost there
Looking into Hyperlinks now

Then I will write tests and it should be good to go

**Reply by dolanmiu** [reply]
Almost done, works really well. Covered by a bunch of tests, 99% coverage.

Will merge shortly, and release v8

**Reply by martijn10kb** [reply]
@dolanmiu Nice!

We already have the branch you're working on in a branch checked out and configured..
We are running in this error log though, not quite sure where the problem is.
```log
Failed <ref *1> Error: Attribute without value
Line: 133
Column: 341
Char: >
    at R (/reports/node_modules/docx/build/index.js:2:141296)
    at k (/reports/node_modules/docx/build/index.js:2:141672)
    at s.write (/reports/node_modules/docx/build/index.js:2:132374)
    at e.exports (/reports/node_modules/docx/build/index.js:2:200441)
    at Fa (/reports/node_modules/docx/build/index.js:2:331706)
    at i (/reports/node_modules/docx/build/index.js:2:336348)
    at i.next (<anonymous>)
    at o (/reports/node_modules/docx/build/index.js:2:338428) {
  note: [Circular *1]
}
```
It might be that we have a key that doesn't have a value in the patches object. But which one from the >100 values..

**Reply by martijn10kb** [reply]
And I'm trying to run the demo 85. But I receive this other error; some other seem to work fine.

```
prompt: number:  85
Running demo 85: 85-template-document.ts

> docx@7.8.2 ts-node /reports/node_modules/docx
> ts-node --skip-project "./demo/85-template-document.ts"

/reports/node_modules/docx/node_modules/ts-node/src/index.ts:859
    return new TSError(diagnosticText, diagnosticCodes, diagnostics);
           ^
TSError: ⨯ Unable to compile TypeScript:
src/file/paragraph/formatting/alignment.ts(4,53): error TS2307: Cannot find module '@file/xml-components' or its corresponding type declarations.
src/file/paragraph/formatting/alignment.ts(58,14): error TS2339: Property 'root' does not exist on type 'Alignment'.
src/file/paragraph/formatting/alignment.ts(58,48): error TS2554: Expected 0 arguments, but got 1.

    at createTSError (/reports/node_modules/docx/node_modules/ts-node/src/index.ts:859:12)
    at reportTSError (/reports/node_modules/docx/node_modules/ts-node/src/index.ts:863:19)
    at getOutput (/reports/node_modules/docx/node_modules/ts-node/src/index.ts:1077:36)
    at Object.compile (/reports/node_modules/docx/node_modules/ts-node/src/index.ts:1433:41)
    at Module.m._compile (/reports/node_modules/docx/node_modules/ts-node/src/index.ts:1617:30)
    at Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Object.require.extensions.<computed> [as .ts] (/reports/node_modules/docx/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19) {
  diagnosticCodes: [ 2307, 2339, 2554 ]
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! docx@7.8.2 ts-node: `ts-node --skip-project "./demo/85-template-document.ts"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the docx@7.8.2 ts-node script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2023-03-16T14_01_08_564Z-debug.log
Something went wrong with the demo
```

**Reply by dolanmiu** [reply]
Yes, it's because I have temporarily imported from `/src` rather than `/build` in that particular demo. Will fix

**Reply by dolanmiu** [reply]
Released as part of `v8.0.0`! Enjoy

**Reply by martijn10kb** [reply]
Thank you very much we have checked it out! And appreciate your work a lot!

Though we are still running into this issue:

> @dolanmiu Nice!
> 
> We already have the branch you're working on in a branch checked out and configured.. We are running in this error log though, not quite sure where the problem is.
> 
> ```
> Failed <ref *1> Error: Attribute without value
> Line: 133
> Column: 341
> Char: >
>     at R (/reports/node_modules/docx/build/index.js:2:141296)
>     at k (/reports/node_modules/docx/build/index.js:2:141672)
>     at s.write (/reports/node_modules/docx/build/index.js:2:132374)
>     at e.exports (/reports/node_modules/docx/build/index.js:2:200441)
>     at Fa (/reports/node_modules/docx/build/index.js:2:331706)
>     at i (/reports/node_modules/docx/build/index.js:2:336348)
>     at i.next (<anonymous>)
>     at o (/reports/node_modules/docx/build/index.js:2:338428) {
>   note: [Circular *1]
> }
> ```
> 
> It might be that we have a key that doesn't have a value in the patches object. But which one from the >100 values..

**Reply by dolanmiu** [reply]
Interesting, could you minimise the reproduction steps as much as possible? then it could be easier for me to find the root cause

**Reply by martijn10kb** [reply]
Yes, I will try to do that tomorrow for you in one of your sample docx.. ..I do hope I can reproduce it in that document..

**Reply by martijn10kb** [reply]
@dolanmiu Hmm can't really isolate it. No matter how I modify your document it still succeeds. And no matter how I edit my document it still fails... 
This document is now completely empty but it still fails if I run it as input document in demo 85.
Could you maybe have a look it at? Maybe you've better luck/skills with docx's internals..

[e738cf8d-3070-4387-b043-b34ffa676039 (2).docx](https://github.com/dolanmiu/docx/files/11060355/e738cf8d-3070-4387-b043-b34ffa676039.2.docx)

**Reply by martijn10kb** [reply]
@dolanmiu I just came to learn that it may have been fixed in 8.0.1 trough the PR from Ewout :)

**Reply by dolanmiu** [reply]
Great! good to know. 

Let me know if there are any other issues

**Reply by martijn10kb** [reply]
@dolanmiu Alright this works really cool!
For the next step. It would be really great if we had some kind of function support. Like i can have in word the following..

```
{{ generateTableFromData('employeeCollection', (a) => a.isNotFired, ['first_name', 'last_name', 'age']  ) }}
{{ generateTableFromData('documentCollection', (a) => a.type === 'foo', ['path', 'title', 'description']  ) }}
```

And then in your code you have something like this:

```ts
const { data } = graphqlQuery()

await patchDocument(data, {
  patches: [
    ...,
    generateTableFromData: (collection: string, filter: Fn, columns: Array<string>) => {
      return new table(children: [
        new row(isHeader: true, children: column.map(i => t(i)) ),
        ...data[collection].filter(filter).map((item) =>{
          return new row(children: columns.map( ...generate column per row..)
        }),
      ])
    },
    ...,
  ]
});
```

Although this would be really great I think.. ..it still might not 100% work for our needs I think..
Maybe @ewoutkleinsmann has another idea?

**Comment by DharyaSoni** [comment]
Please Suggest, How can I show rich text as html tag formatted string in Documents.
Eg. "<p> Text to Show</p>".

**Reply by martijn10kb** [reply]
See the demo case 85.
Besides that you will need your own rich text parser to docx mapper.
I cant share the code we used 1 on 1, but we use `himalaya` to parse the rich text to json. Then we have a recursive function that generates all the elements like `TextRun`, `ExternalHyperlink`, `Paragraph`, etc.

**Reply by dolanmiu** [reply]
Interesting approach!

**Reply by DharyaSoni** [reply]
> See the demo case 85. Besides that you will need your own rich text parser to docx mapper. I cant share the code we used 1 on 1, but we use `himalaya` to parse the rich text to json. Then we have a recursive function that generates all the elements like `TextRun`, `ExternalHyperlink`, `Paragraph`, etc.

Could you please share that recursive method that generate all elements like text run, and all

**Reply by martijn10kb** [reply]
```ts
import { parse } from 'himalaya';

export default html = (html: string, root = true) => {
  const elements = parse(html);
  return mapToDocx(elements, root);
};

const mapToDocx = (elements: Element[], styles = {}, root: boolean): any[] => {
  return elements.flatMap((element) => toDocx(element, root));
};

export const toDocx = (
  { type, tagName, attributes, content, children }: Element,
  extraStyles: object,
  root: boolean,
) => {

  ...here have your code..
  fi. if type === 'text' return TextRun({...extraStyles, children: content})
  if tagname is bold or b return mapToDocx(children, { bold: true }, root);
  if tagname img return imageRun
  etcetc..
  
  return [];
};

```
